adminapi/zsh-completion/adminapi

```diff
@@ -0,0 +1,62 @@
+# Copyright © 2021, InnoGames GmbH
+# Query for all hosts from serveradmin API
+# This file should be put to the directory in ${FPATH},
+#   Arch Linux, Fedora: /usr/share/zsh/site-functions
+#   Debian: /usr/share/zsh/vendor-functions
+#
+# Configuration:
+# _SERVERADMIN_HOSTS_ENABLED: if the completion should be enabled
+# _SERVERADMIN_HOSTS_QUERY: customize hosts list, string
+# _SERVERADMIN_HOSTS_TIMEOUT: customize serveradmin query timeout, integer, seconds
+# _SERVERADMIN_HOSTS_UPDATE: invalidate cache interval, integer, minutes
+# _SERVERADMIN_HOSTS_ZSTYLE: change autocompletion behavior for hosts, bool
+function _serveradmin_hosts() {
+  # Not enable if adminapi is not found
+  command -v adminapi 1>/dev/null || return
+  local _query=${_SERVERADMIN_HOSTS_QUERY:-'servertype=any(blade_enclosure 
+    floating_address hardware hw_internalrouter hw_loadbalancer hypervisor 
+    loadbalancer natbox switch vm vm_external vpn)'}
+  local _hosts_cache="${HOME}/.ssh/serveradmin_hosts_cache"
+  local _tmp="${_hosts_cache}.tmp"
+  local _lock="${_hosts_cache}.lock"
+  local _invalidate_minutes=${_SERVERADMIN_HOSTS_UPDATE:-1}
+  local _timeout="${_SERVERADMIN_HOSTS_TIMEOUT:-8}s"
+  if [ "${_hosts_cache}"(Nmm-${_invalidate_minutes}) ]; then
+    # File exists and newer then $_invalidate_minutes minutes
+    # Use "<" as zsh internal `cat` for not invoking extra process
+    < "${_hosts_cache}"
+    return
+  fi
+
+  if [ -f "${_hosts_cache}" ]; then
+    # File exists and older then $_invalidate_minutes minutes
+    # First return list, then update it in background
+    < "${_hosts_cache}"
+    if   flock -n "${_lock}" true && [ -s "${_tmp}" ]; then
+      # If lockfile not locked and new temp cache already exists and not empty
+      mv "${_tmp}" "${_hosts_cache}"
+    elif flock -n "${_lock}" true; then
+      # It shouldn't hang forever in case of problems with serveradmin
+      flock -n "${_lock}" timeout $_timeout \
+        adminapi $_query 2>/dev/null > "${_tmp}" &
+    fi
+    return
+  fi
+
+  # If file doesn't exists, then waiting for getting fresh
+  timeout $_timeout adminapi $_query 2>/dev/null | tee "${_hosts_cache}"
```

Comment:
Out of curiosity: Have you tested how this ends the HTTP connection if we run into the timeout ? Will this be visible as 499 on the backend side ?

Reply:
Yes, this is ended up as 499. But in our case, 8sec is sufficient timeout

Reply:
Alright - sounds fair. Yes 8 seconds should be sufficient for simple Queries even for our size of database. This might just explain some 499 requests we have so it is good to know this can happen.